File: kasandra-crawler/src/main/scala/com/jaitlapps/kasandra/crawler/parsers/VkWallParser.scala
package com.jaitlapps.kasandra.crawler.parsers

import com.fasterxml.jackson.databind.ObjectMapper
import com.jaitlapps.kasandra.crawler.models.CrawledVkUrl
import collection.JavaConverters._

object VkWallParser {

  private val regex = "(https?)://[^\\s/$.?#].[^\\s]*".r

  def parseJson(data: String): Set[CrawledVkUrl] = {
    val mapper = new ObjectMapper()
    val tree = mapper.readTree(data)
    val nodes = tree.get("response").get("items").elements().asScala.toSet

    nodes.flatMap(jNode => {
      val date = jNode.get("date").asLong() * 1000
      val post = jNode.get("text").asText()

      val urls = parseUrls(post).map(u => CrawledVkUrl(u, date))

      if (jNode.has("attachments")) {
        val link = jNode.get("attachments").asScala.toSeq.find(att => att.get("type").asText() == "link")
        if(link.isDefined) {
          val urlLink = link.get.get("link").get("url").asText()
          urls :+ CrawledVkUrl(urlLink, date)
        } else {
          urls
        }
      } else {
        urls
      }
    })
  }

  def parseWallSize(data: String): Int = {
    val mapper = new ObjectMapper()
    val tree = mapper.readTree(data)

    tree.get("response").get("count").asInt()
  }

  private def parseUrls(text: String): Seq[String] = regex.findAllIn(text).toSeq
}

File: kasandra_scrapy/kasandra_scrapy/parsers/vk_wall_parser.py
# -*- coding: utf-8 -*-
import json
import re
from .vk_wall_item import VkWallItem


class VkWallParser(object):
    def __init__(self):
        self.__regex = "(https?://[^\\s/$.?#].[^\\s]*)"

    def parse(self, json_data):
        data = json.loads(json_data)
        items = data['response']['items']

        result = []

        for item in items:
            date = item['date']
            text = item['text']

            urls = set(re.findall(self.__regex, text))

            try:
                for attachment in item['attachments']:
                    if attachment['type'] == 'link':
                        urls.add(attachment['link']['url'])
            except:
                pass

            for url in urls:
                wall_item = VkWallItem(url, date)
                result.append(wall_item)

        return result
